File: dynamic_zero_export/runtime/source-detection.ts
import { type DynamicZeroExportSiteConfig, type GeneratorConfig } from '../schema/site-config.types';

export type SourceState = 'GRID' | 'GENERATOR' | 'NONE' | 'AMBIGUOUS';

export type RealMeterSample = {
  kw: number;
  stale: boolean;
  sourceHint?: SourceState;
};

export function detectSource(sample: RealMeterSample): SourceState {
  if (sample.stale) return 'AMBIGUOUS';
  if (sample.sourceHint) return sample.sourceHint;
  if (sample.kw > 0) return 'GENERATOR';
  if (sample.kw < 0) return 'GRID';
  return 'NONE';
}

export function generatorMinimumLoadKw(
  config: DynamicZeroExportSiteConfig,
  generators: GeneratorConfig[],
): number {
  return generators.reduce((sum, generator) => {
    const pct =
      generator.type === 'gas'
        ? config.policy.gasMinimumLoadPct
        : config.policy.dieselMinimumLoadPct;
    return sum + (generator.ratingKw * pct) / 100;
  }, 0);
}

export function normalizeMeterKw(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.round(value * 1000) / 1000;
}

File: dynamic_zero_export/firmware_core/src/source_detection.cpp
#include "dzx/source_detection.hpp"

namespace dzx {

ActiveSource detectSource(const RealMeterSample& sample) {
  if (sample.stale) return ActiveSource::Ambiguous;
  if (!sample.sourceKnown) return ActiveSource::Ambiguous;
  if (sample.source == "GRID") return ActiveSource::Grid;
  if (sample.source == "GENERATOR") return ActiveSource::Generator;
  if (sample.source == "NONE") return ActiveSource::None;
  return ActiveSource::Ambiguous;
}

double generatorMinimumLoadKw(const DynamicZeroExportSiteConfig& cfg, const std::vector<GeneratorConfig>& generators) {
  double total = 0.0;
  for (const auto& gen : generators) {
    const double pct = gen.type == GeneratorType::Diesel
      ? cfg.policy.dieselMinimumLoadPct
      : cfg.policy.gasMinimumLoadPct;
    total += gen.ratingKw * (pct / 100.0);
  }
  return total;
}

}  // namespace dzx

